File: python/leetcode/shortest-completing-word.py
class Solution:
    def shortestCompletingWord(self, license_plate, words):
        """
        :type licensePlate: str
        :type words: List[str]
        :rtype: str
        """
        lp = self.pull_letters(license_plate)
        min_length = 1000 # max specified by task
        shortest_word = None
        for i in range(len(words)):
            if self.has_match(lp, words[i]):
                if len(words[i]) < min_length:
                    min_length = len(words[i])
                    shortest_word = words[i]
        return shortest_word
    
    def pull_letters(self, license_plate):
        letters_only = []
        for c in license_plate.lower():
            if c.isalpha():
                letters_only.append(c)
        return letters_only
    
    def has_match(self, lp, word):
        candidates = lp.copy()
        for i in range(len(word)):
            try:
                char_index = candidates.index(word[i])
                del candidates[char_index]
            except:
                pass
            if len(candidates) == 0:
                return True
        return False

File: java/leetcode/shortest-completing-word.java
class Solution {
    public String shortestCompletingWord(String licensePlate, String[] words) {
        char[] lp = pullLetters(licensePlate);
        int minLen = 1000;
        String shortestWord = "";
        for (int i = 0; i < words.length; i++) {
            if (hasMatch(lp, words[i])) {
                if (words[i].length() < minLen) {
                    minLen = words[i].length();
                    shortestWord = words[i]; 
                }
            }
        }
        return shortestWord;
    }
    
    private char[] pullLetters(String licensePlate) {
        return licensePlate.toLowerCase().replaceAll("[\\d|\\s]?", "").toCharArray();
    }
    
    private boolean hasMatch(char[] lp, String word) {
        // checks if all letters contained in lp char array
        int total = lp.length;
        char[] lpc = lp.clone();
        for (int i = 0; i < word.length(); i++) {
            int charIndex = indexOf(lpc, word.charAt(i));
            if (charIndex != -1) {
                lpc[charIndex] = ' ';
                total--;
            }
            if (total == 0) { return true; }
        }
        return false;
    }
    
    private int indexOf(char[] lp, char search) {
        for (int i = 0; i < lp.length; i++) {
            if (lp[i] == search) { return i; }
        }
        
        return -1;
    }
}
